### PR TITLE
Store and fetch comments using Datastore

### PR DESCRIPTION
**Pull request by vaishnavi-janardhan**

### CODE REVIEW

**Comment by vaishnavi-janardhan** [comment]
## Changes
- Modify DataServlet.java to respond with a hard-coded message
- Use fetch() to request message from Servlet and add it to the DOM
- Fetch hard-coded comments in JSON format from server and display

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -14,19 +14,58 @@
...
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String text = request.getParameter("text-input");
+
+    if (!text.equals("")) {
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("message",text);
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: do not use magic strings, use a constants file to reference all magic strings

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.google.sps.data;
+
+public final class Comment {
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:5]: Use Project Lombok. Refer to Sai's PR posted on chat for more details

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -14,19 +14,59 @@
...
+
+    if (StringUtils.isNotEmpty(text)) {
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("message",text);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(commentEntity);
+    }
+    
+    response.sendRedirect("/index.html");
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: why redirect? Just return a response.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,19 +14,59 @@
...
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment");
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Same as above for Magic Strings

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,19 +14,59 @@
...
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String message = (String) entity.getProperty("message");
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Same as above for Magic Strings

**Review by skashish-sps — CHANGES_REQUESTED** [review]
As a practise, always add the deployed link to check for possible errors

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: A redirect is needed to go back to the homepage, else it stays at /data. I can't return a response as doPost() returns void.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Added Constants.java to store all magic strings.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Referenced from Constants.java

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Referenced from Constants.java

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:5]: Used Lombok annotations for auto-generated constructors.

**Comment by vaishnavi-janardhan** [comment]
[Portfolio link](https://vjanardhan-sps-summer20.appspot.com/)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (skashish-sps) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It does not return the void response. The Http response is returned as a reference, which is the argument. Check for the status code on the client side, and then add the comment to the comment container

Comment (skashish-sps) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Also add some validation on the client side. I added a single space as a comment.

Comment (skashish-sps) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: An empty comment does not return an error. It just redirects to index.html

File: portfolio/src/main/java/com/google/sps/Constants.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.google.sps;
+
+public final class Constants {
+
+  private Constants() {
```
Comment (skashish-sps) [inline portfolio/src/main/java/com/google/sps/Constants.java:5]: I want to understand, why you did this?
I mean, why did you feel the need for this class to not be instantiated

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Added client and server side validation.
Also removed redirection and loaded comments based on the response.

**Comment by vaishnavi-janardhan** [comment]
@skashish-sps I have added validation checks and removed redirection.
Can you please check if this is the expected behaviour?
[Portfolio link](https://vjanardhan-sps-summer20.appspot.com)

**Review by skashish-sps — APPROVED** [review]
LGTM, just answer the question in Constants.java

File: portfolio/src/main/java/com/google/sps/Constants.java (hunk 6 above)
Comment (vaishnavi-janardhan) [reply portfolio/src/main/java/com/google/sps/Constants.java:5]: I thought that the Constants class is just a utility class and wanted to ensure that it is non-instantiable.
I couldn't think of why someone would want to instantiate the Constants class.
I remember reading this in a Stack Overflow post.

**Pull request merged**